### PR TITLE
Guard file settings readiness on file settings support

### PR DESCRIPTION
**Pull request by rjernst**
Consistency of file settings is an important invariant. However, when upgrading from Elasticsearch versions before file settings existed, cluster state will not yet have the file settings metadata. If the first node upgraded is not the master node, new nodes will never become ready while they wait for file settings metadata to exist.

This commit adds a node feature for file settings to guard waiting on file settings for readiness. Although file settings has existed since 8.4, the feature is not a historical feature because historical features are not applied to cluster state that readiness checks. In this case it is not needed since upgrading from 8.4+ will already contain file settings metadata.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/readiness/ReadinessService.java [hunk 1]
```diff
@@ -277,7 +279,20 @@ private boolean isMasterElected(ClusterState clusterState) {
...
+            // readiness until the whole cluster supports file settings.
+            return supportsFileSettings(clusterState) == false;
+        } else {
+            return fileSettingsMetadata.version().equals(ReservedStateMetadata.NO_VERSION) == false;
+        }
+    }
+
+    @SuppressForbidden(reason = "need to check file settings support on exact cluster state")
+    private static boolean supportsFileSettings(ClusterState clusterState) {
+        return clusterState.clusterFeatures().clusterHasFeature(FileSettingsFeatures.FILE_SETTINGS_SUPPORTED);
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/readiness/ReadinessService.java:295]: This is what `FeatureService.clusterHasFeature` does, no?

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsFeatures.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import org.elasticsearch.features.FeatureSpecification;
+import org.elasticsearch.features.NodeFeature;
+
+import java.util.Set;
+
+public class FileSettingsFeatures implements FeatureSpecification {
+
+    // Although file settings were supported starting in 8.4.0, this is really about whether file settings
+    // are used in readiness.
+    public static final NodeFeature FILE_SETTINGS_SUPPORTED = new NodeFeature("file_settings");
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsFeatures.java:20]: There's normally a scoping section to the feature id. So here, something like `readiness.considers_file_settings`

File: server/src/test/java/org/elasticsearch/readiness/ReadinessServiceTests.java [hunk 3]
```diff
@@ -324,4 +307,45 @@ public void testFileSettingsUpdateError() throws Exception {
...
+
+    public void testFileSettingsMixedCluster() throws Exception {
+        readinessService.start();
+
+        // initially the service isn't ready because initial cluster state has not been applied yet
+        assertFalse(readinessService.ready());
+
+        ClusterState noFileSettingsState = ClusterState.builder(noFileSettingsState())
+            // the master node is upgraded to support file settings, but existing node2 is not
+            .nodeFeatures(Map.of(httpTransport.node.getId(), Set.of(FileSettingsFeatures.FILE_SETTINGS_SUPPORTED.id())))
```
Comment (thecoop) [inline server/src/test/java/org/elasticsearch/readiness/ReadinessServiceTests.java:319]: We've already got `nodeFeatures` static var for the set

File: server/src/main/java/org/elasticsearch/readiness/ReadinessService.java [hunk 4]
```diff
@@ -277,7 +279,20 @@ private boolean isMasterElected(ClusterState clusterState) {
     // protected to allow mock service to override
     protected boolean areFileSettingsApplied(ClusterState clusterState) {
         ReservedStateMetadata fileSettingsMetadata = clusterState.metadata().reservedStateMetadata().get(FileSettingsService.NAMESPACE);
-        return fileSettingsMetadata != null && fileSettingsMetadata.version().equals(ReservedStateMetadata.NO_VERSION) == false;
+        if (fileSettingsMetadata == null) {
+            // In order to block readiness on file settings being applied, we need to know that the master node has written an initial
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/readiness/ReadinessService.java:283]: Maybe worth a comment re-iterating that this is only if the metadata is not found, indicating it is <8.4 anyway

**Review by thecoop — APPROVED** [review]
OK, couple of minor comments to check

File: server/src/main/java/org/elasticsearch/readiness/ReadinessService.java (hunk 1 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/readiness/ReadinessService.java:295]: Historical features are added by the FeatureService, so we can't check them here against a specific cluster state

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsFeatures.java (hunk 2 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsFeatures.java:20]: I see the "supported" context in some examples. I also don't like "considers".

**Pull request merged**